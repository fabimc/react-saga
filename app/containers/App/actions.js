/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_REPOS,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_ERROR,
  LOAD_RATES,
  LOAD_RATES_SUCCESS,
  LOAD_RATES_ERROR
} from './constants'

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRepos () {
  return { type: LOAD_REPOS }
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposLoaded (repos, username) {
  return { type: LOAD_REPOS_SUCCESS, repos, username }
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function repoLoadingError (error) {
  return { type: LOAD_REPOS_ERROR, error }
}

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRates () {
  return { type: LOAD_RATES }
}

/**
 * Dispatched when the rates are loaded by the request saga
 *
 * @param  {array} rates The rates data
 * @param  {string} base The current base
 *
 * @return {object}      An action object with a type of LOAD_RATES_SUCCESS passing the rates
 */
export function ratesLoaded (rates, base) {
  return { type: LOAD_RATES_SUCCESS, rates, base }
}

/**
 * Dispatched when loading the rates fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function ratesLoadingError (error) {
  return { type: LOAD_RATES_ERROR, error }
}
