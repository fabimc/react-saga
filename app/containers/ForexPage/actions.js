/*
 * Forex Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_BASE,
  LOAD_RATES,
  LOAD_RATES_SUCCESS,
  LOAD_RATES_ERROR
} from './constants'

/**
 * Changes the input field of the form
 *
 * @param  {name} name The new text of the input field
 *
 * @return {object}    An action object with a type of CHANGE_BASE
 */
export function changeBase (base) {
  return { type: CHANGE_BASE, base }
}

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRates () {
  return { type: LOAD_RATES }
}

/**
 * Dispatched when the rates are loaded by the request saga
 *
 * @param  {array} rates The rates data
 * @param  {string} base The current base
 *
 * @return {object}      An action object with a type of LOAD_RATES_SUCCESS passing the rates
 */
export function ratesLoaded (rates, base) {
  return { type: LOAD_RATES_SUCCESS, rates, base }
}

/**
 * Dispatched when loading the rates fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function ratesLoadingError (error) {
  return { type: LOAD_RATES_ERROR, error }
}
