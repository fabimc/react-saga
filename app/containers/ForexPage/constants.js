/*
 * ForexConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const CHANGE_BASE = 'boilerplate/Forex/CHANGE_BASE'
export const LOAD_RATES = 'boilerplate/Forex/LOAD_RATES'
export const LOAD_RATES_SUCCESS = 'boilerplate/Forex/LOAD_RATES_SUCCESS'
export const LOAD_RATES_ERROR = 'boilerplate/Forex/LOAD_RATES_ERROR'
