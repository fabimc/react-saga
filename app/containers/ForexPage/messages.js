/*
 * ForexPage Messages
 *
 * This contains all the text for the ForexPage component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  startProjectHeader: {
    id: 'boilerplate.containers.ForexPage.start_project.header',
    defaultMessage: 'Start your next react project in seconds',
  },
  startProjectMessage: {
    id: 'boilerplate.containers.ForexPage.start_project.message',
    defaultMessage: 'A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices',
  },
  trymeHeader: {
    id: 'boilerplate.containers.ForexPage.tryme.header',
    defaultMessage: 'Try me!',
  },
  trymeMessage: {
    id: 'boilerplate.containers.ForexPage.tryme.message',
    defaultMessage: 'Show Github ratesitories by',
  },
  trymeAtPrefix: {
    id: 'boilerplate.containers.ForexPage.tryme.atPrefix',
    defaultMessage: '@',
  },
});
