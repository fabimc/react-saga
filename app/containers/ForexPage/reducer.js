/*
 * ForereignexReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */
import { fromJS } from 'immutable'

import {
  CHANGE_BASE,
  LOAD_RATES_SUCCESS,
  LOAD_RATES,
  LOAD_RATES_ERROR
} from './constants'

// The initial state of the App
const initialState = fromJS({ base: '', ratesData: { rates: false } })

function forexReducer (state = initialState, action) {
  switch (action.type) {
    case CHANGE_BASE:
      // Delete prefixed '@' from the github base
      return state.set('base', action.base.replace(/@/gi, ''))
    case LOAD_RATES:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn([ 'ratesData', 'rates' ], false)
    case LOAD_RATES_SUCCESS:
      return state
        .setIn([ 'ratesData', 'rates' ], action.rates.rates)
        .setIn([ 'ratesData', 'base' ], action.rates.base)
        .setIn([ 'ratesData', 'date' ], action.rates.date)
        .set('loading', false)
    case LOAD_RATES_ERROR:
      return state.set('error', action.error).set('loading', false)
    default:
      return state
  }
}

export default forexReducer
