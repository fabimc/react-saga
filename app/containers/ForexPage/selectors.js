/**
 * Forexpage selectors
 */

import { createSelector } from 'reselect'

const selectForex = state => state.get('forex')

const makeSelectBase = () =>
  createSelector(selectForex, forexState => forexState.get('base'))

export { selectForex, makeSelectBase }
