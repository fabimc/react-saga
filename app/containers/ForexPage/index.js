/*
 * ForexPage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react'
import Helmet from 'react-helmet'
import { FormattedMessage } from 'react-intl'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import {
  makeSelectRates,
  makeSelectLoading,
  makeSelectError
} from 'containers/App/selectors'
import H2 from 'components/H2'
// import ReposList from 'components/ReposList'
import AtPrefix from './AtPrefix'
import CenteredSection from './CenteredSection'
import Form from './Form'
import Input from './Input'
import Section from './Section'
import messages from './messages'
import { loadRates } from '../App/actions'
import { changeBase } from './actions'
import { makeSelectBase } from './selectors'

export class ForexPage extends React.PureComponent {
  // eslint-disable-line react/prefer-stateless-function
  /**
   * when initial state base is not null, submit the form to load rates
   */
  componentDidMount () {
    if (this.props.base && this.props.base.trim().length > 0) {
      this.props.onSubmitForm()
    }
  }

  render () {
    const { loading, error, rates } = this.props
    const ratesListProps = { loading, error, rates }

    return (
      <article>
        <Helmet
          title='Forex Page'
          meta={
            [
              {
                name: 'description',
                content: 'A React.js Boilerplate application homepage'
              }
            ]
          }
        />
        <div>
          <CenteredSection>
            <H2>
              <FormattedMessage {...messages.startProjectHeader} />
            </H2>
            <p>
              <FormattedMessage {...messages.startProjectMessage} />
            </p>
          </CenteredSection>
          <Section>
            <H2>
              <FormattedMessage {...messages.trymeHeader} />
            </H2>
            <Form onSubmit={this.props.onSubmitForm}>
              <label htmlFor='base'>
                <FormattedMessage {...messages.trymeMessage} />
                <AtPrefix>
                  <FormattedMessage {...messages.trymeAtPrefix} />
                </AtPrefix>
                <Input
                  id='base'
                  type='text'
                  placeholder='USD'
                  value={this.props.base}
                  onChange={this.props.onChangeBase}
                />
              </label>
            </Form>
            <p> here goes the list </p>
          </Section>
        </div>
      </article>
    )
  }
}

ForexPage.propTypes = {
  loading: React.PropTypes.bool,
  error: React.PropTypes.oneOfType([
    React.PropTypes.object,
    React.PropTypes.bool
  ]),
  rates: React.PropTypes.oneOfType([
    React.PropTypes.object,
    React.PropTypes.bool
  ]),
  onSubmitForm: React.PropTypes.func,
  base: React.PropTypes.string,
  onChangeBase: React.PropTypes.func
}

export function mapDispatchToProps (dispatch) {
  return {
    onChangeBase: evt => dispatch(changeBase(evt.target.value)),
    onSubmitForm: evt => {
      if (evt !== undefined && evt.preventDefault) evt.preventDefault()
      dispatch(loadRates())
    }
  }
}

const mapStateToProps = createStructuredSelector({
  rates: makeSelectRates(),
  base: makeSelectBase(),
  loading: makeSelectLoading(),
  error: makeSelectError()
})

// Wrap the component to inject dispatch and state into it
export default connect(mapStateToProps, mapDispatchToProps)(ForexPage)
