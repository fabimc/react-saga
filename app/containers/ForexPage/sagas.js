/**
 * Gets the ratesitories of the user from Github
 */

import {
  take,
  call,
  put,
  select,
  cancel,
  takeLatest
} from 'redux-saga/effects'
import { LOCATION_CHANGE } from 'react-router-redux'
import { LOAD_RATES } from 'containers/ForexPage/constants'
import { ratesLoadingError, ratesLoaded } from 'containers/ForexPage/actions'

import request from 'utils/request'
import { makeSelectBase } from 'containers/ForexPage/selectors'

/**
 * Github rates request/response handler
 */
export function* getRates () {
  // Select base from store
  const base = yield select(makeSelectBase())
  const requestURL = `http://api.fixer.io/latest?base=${base}`

  try {
    // Call our request helper (see 'utils/request')
    const rates = yield call(request, requestURL)
    yield put(ratesLoaded(rates, base))
  } catch (err) {
    yield put(ratesLoadingError(err))
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* ratesData () {
  // Watches for LOAD_RATES actions and calls getRates when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_RATES, getRates)

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE)
  yield cancel(watcher)
}

// Bootstrap sagas
export default [ ratesData ]
