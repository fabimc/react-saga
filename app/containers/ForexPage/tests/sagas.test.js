/**
 * Tests for ForexPage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects'
import { createMockTask } from 'redux-saga/lib/utils'

import { LOCATION_CHANGE } from 'react-router-redux'

import { LOAD_RATES } from 'containers/ForexPage/constants'
import { ratesLoaded, ratesLoadingError } from 'containers/ForexPage/actions'

import { getRates, ratesData } from '../sagas'

const base = 'USD'

/* eslint-disable redux-saga/yield-effects */
describe('getRates Saga', () => {
  let getRatesGenerator

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getRatesGenerator = getRates()

    const selectDescriptor = getRatesGenerator.next().value
    expect(selectDescriptor).toMatchSnapshot()

    const callDescriptor = getRatesGenerator.next(base).value
    expect(callDescriptor).toMatchSnapshot()
  })

  it(
    'should dispatch the ratesLoaded action if it requests the data successfully',
    () => {
      const response = [ { name: 'First repo' }, { name: 'Second repo' } ]
      const putDescriptor = getRatesGenerator.next(response).value
      expect(putDescriptor).toEqual(put(ratesLoaded(response, base)))
    }
  )

  it('should call the ratesLoadingError action if the response errors', () => {
    const response = new Error('Some error')
    const putDescriptor = getRatesGenerator.throw(response).value
    expect(putDescriptor).toEqual(put(ratesLoadingError(response)))
  })
})

describe('ratesDataSaga Saga', () => {
  const ratesDataSaga = ratesData()
  const mockedTask = createMockTask()

  it('should start task to watch for LOAD_RATES action', () => {
    const takeLatestDescriptor = ratesDataSaga.next().value
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_RATES, getRates))
  })

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = ratesDataSaga.next(mockedTask).value
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE))
  })

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = ratesDataSaga.next().value
    expect(cancelDescriptor).toEqual(cancel(mockedTask))
  })
})
