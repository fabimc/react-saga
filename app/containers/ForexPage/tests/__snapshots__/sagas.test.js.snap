exports[`getRates Saga should call the ratesLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getRates Saga should call the ratesLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://api.fixer.io/latest?base=mxstbr",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getRates Saga should dispatch the ratesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getRates Saga should dispatch the ratesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://api.fixer.io/latest?base=mxstbr",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
